File: src/bigtube/core/converter_history.py
import os
import threading
import time

from gi.repository import GLib

from .json_store import load_json, save_json
from .logger import get_logger

# Module logger
logger = get_logger(__name__)


class ConverterHistoryManager:
    """
    Manages the persistence of conversion history.
    Stores data in a JSON file within the user's config directory.
    """

    # We use the same directory logic as ConfigManager
    _CONFIG_DIR = os.path.join(GLib.get_user_config_dir(), "bigtube")
    _FILE_PATH = os.path.join(_CONFIG_DIR, "converter_history.json")

    # Maximum number of items to keep in conversion history
    MAX_HISTORY_SIZE = 50
    _lock = threading.RLock()
    _cache: list | None = None
    _pending_save = False
    _DEBOUNCE_DELAY = 2.0

    @classmethod
    def load(cls) -> list:
        """
        Reads the history from disk.
        Returns an empty list if the file does not exist or is corrupted.
        """
        with cls._lock:
            if cls._cache is not None:
                return cls._cache.copy()

        data = load_json(cls._FILE_PATH, [])
        with cls._lock:
            cls._cache = data
        return data.copy()

    @classmethod
    def _save_to_disk(cls):
        with cls._lock:
            if cls._cache is None:
                logger.debug("Skipping converter history save: cache not loaded.")
                return
            cls._pending_save = False
            data_to_save = cls._cache.copy()

        save_json(cls._FILE_PATH, data_to_save, indent=2)

    @classmethod
    def _debounced_save_callback(cls):
        cls._save_to_disk()
        return False

    @classmethod
    def _schedule_save(cls):
        with cls._lock:
            if cls._pending_save:
                return
            cls._pending_save = True

        GLib.timeout_add(int(cls._DEBOUNCE_DELAY * 1000), cls._debounced_save_callback)

    @classmethod
    def save(cls, items: list):
        """Updates cache and schedules a debounced disk write."""
        with cls._lock:
            cls._cache = items.copy()
        cls._schedule_save()

    @classmethod
    def save_immediate(cls, items: list):
        """Updates cache and writes immediately."""
        with cls._lock:
            cls._cache = items.copy()
            cls._pending_save = False
        cls._save_to_disk()

    @classmethod
    def add_entry(cls, source_path: str, output_path: str, format_id: str):
        """
        Adds or updates a conversion entry.
        If the same source and format exist, it updates the timestamp and output path.
        """
        with cls._lock:
            history = cls.load()

            # Remove existing entry for the same conversion to avoid duplicates
            # and keep the latest output for that format at the top.
            history = [
                item
                for item in history
                if not (item.get("source") == source_path and item.get("format") == format_id)
            ]

            new_item = {
                "source": source_path,
                "output": output_path,
                "format": format_id,
                "timestamp": time.time(),
            }

            # Insert at the beginning (Newest first)
            history.insert(0, new_item)

            # Limit history size
            history = history[: cls.MAX_HISTORY_SIZE]

            cls.save(history)
            return new_item

    @classmethod
    def remove_entry(cls, source_path: str, format_id: str = None):
        """
        Removes an item from history.
        If format_id is None, removes all formats for that source.
        """
        with cls._lock:
            history = cls.load()
            if format_id:
                new_history = [
                    item
                    for item in history
                    if not (item.get("source") == source_path and item.get("format") == format_id)
                ]
            else:
                new_history = [item for item in history if item.get("source") != source_path]

            if len(new_history) != len(history):
                cls.save_immediate(new_history)
                logger.info(f"Removed converter history entry for: {source_path}")

    @classmethod
    def clear_all(cls):
        """
        Wipes the entire converter history.
        """
        with cls._lock:
            cls.save_immediate([])
            logger.info("All converter history entries cleared")

    @classmethod
    def flush(cls):
        """Forces any pending save to disk immediately."""
        with cls._lock:
            if cls._pending_save and cls._cache is not None:
                cls._pending_save = False
        cls._save_to_disk()

    @classmethod
    def _ensure_dir_exists(cls):
        """Helper to create the directory if missing."""
        if not os.path.exists(cls._CONFIG_DIR):
            try:
                os.makedirs(cls._CONFIG_DIR)
            except OSError:
                pass

File: rust/crates/bigtube-core/src/converter_history.rs
//! Conversion-history persistence. Ported from `core/converter_history.py`.
//! Same cache+debounce model as [`crate::history`], deduping by (source, format).

use std::path::PathBuf;
use std::sync::{Arc, Mutex};
use std::time::Duration;

use serde_json::{json, Value};

use crate::debounce::Debouncer;
use crate::json_store::{load_json, save_json};
use crate::util::now_epoch;

pub const MAX_HISTORY_SIZE: usize = 50;
const DEBOUNCE_DELAY: f64 = 2.0;

pub struct ConverterHistoryManager {
    path: PathBuf,
    cache: Arc<Mutex<Option<Vec<Value>>>>,
    debouncer: Debouncer,
    max_size: usize,
}

impl ConverterHistoryManager {
    pub fn new(path: PathBuf) -> Self {
        Self::with_max(path, MAX_HISTORY_SIZE)
    }

    /// Like `new`, but with an explicit history cap (clamped to at least 1).
    pub fn with_max(path: PathBuf, max_size: usize) -> Self {
        let max_size = max_size.max(1);
        let cache: Arc<Mutex<Option<Vec<Value>>>> = Arc::new(Mutex::new(None));
        let debouncer = {
            let cache = cache.clone();
            let path = path.clone();
            Debouncer::new(Duration::from_secs_f64(DEBOUNCE_DELAY), move || {
                let guard = cache.lock().unwrap();
                if let Some(items) = guard.as_ref() {
                    save_json(&path, items, Some(2));
                }
            })
        };
        Self {
            path,
            cache,
            debouncer,
            max_size,
        }
    }

    pub fn load(&self) -> Vec<Value> {
        let mut guard = self.cache.lock().unwrap();
        if let Some(items) = guard.as_ref() {
            return items.clone();
        }
        let data: Vec<Value> = load_json(&self.path, Vec::new());
        *guard = Some(data.clone());
        data
    }

    pub fn save(&self, items: Vec<Value>) {
        *self.cache.lock().unwrap() = Some(items);
        self.debouncer.touch();
    }

    pub fn save_immediate(&self, items: Vec<Value>) {
        *self.cache.lock().unwrap() = Some(items);
        self.debouncer.flush();
    }

    /// Add/update a conversion entry, deduped by (source, format) (`add_entry`).
    pub fn add_entry(&self, source_path: &str, output_path: &str, format_id: &str) -> Value {
        let history = self.load();
        let mut history: Vec<Value> = history
            .into_iter()
            .filter(|item| {
                !(item.get("source").and_then(Value::as_str) == Some(source_path)
                    && item.get("format").and_then(Value::as_str) == Some(format_id))
            })
            .collect();

        let new_item = json!({
            "source": source_path,
            "output": output_path,
            "format": format_id,
            "timestamp": now_epoch(),
        });
        history.insert(0, new_item.clone());
        history.truncate(self.max_size);
        self.save(history);
        new_item
    }

    /// Remove entries for `source_path`; if `format_id` is `None`, remove all
    /// formats for that source (`remove_entry`). Immediate save.
    pub fn remove_entry(&self, source_path: &str, format_id: Option<&str>) {
        let history = self.load();
        let original = history.len();
        let new_history: Vec<Value> = history
            .into_iter()
            .filter(|item| {
                let same_source = item.get("source").and_then(Value::as_str) == Some(source_path);
                match format_id {
                    Some(fmt) => {
                        !(same_source && item.get("format").and_then(Value::as_str) == Some(fmt))
                    }
                    None => !same_source,
                }
            })
            .collect();
        if new_history.len() != original {
            self.save_immediate(new_history);
            tracing::info!("Removed converter history entry for: {source_path}");
        }
    }

    pub fn clear_all(&self) {
        self.save_immediate(Vec::new());
    }

    pub fn flush(&self) {
        self.debouncer.flush();
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn mgr() -> (tempfile::TempDir, ConverterHistoryManager) {
        let dir = tempfile::tempdir().unwrap();
        let m = ConverterHistoryManager::new(dir.path().join("converter_history.json"));
        (dir, m)
    }

    #[test]
    fn add_dedupes_same_source_and_format() {
        let (_d, m) = mgr();
        m.add_entry("/a.mkv", "/a.mp4", "mp4");
        m.add_entry("/a.mkv", "/a2.mp4", "mp4"); // same source+format -> replaces
        m.flush();
        let h = m.load();
        assert_eq!(h.len(), 1);
        assert_eq!(h[0]["output"], json!("/a2.mp4"));
    }

    #[test]
    fn remove_all_formats_for_source() {
        let (_d, m) = mgr();
        m.add_entry("/a.mkv", "/a.mp4", "mp4");
        m.add_entry("/a.mkv", "/a.webm", "webm");
        m.flush();
        assert_eq!(m.load().len(), 2);
        m.remove_entry("/a.mkv", None);
        assert!(m.load().is_empty());
    }
}
